File: Paw_Track-Anshu_IN26011396/src/services/visitService.ts
import { dbContext } from '../data/dbContext';
import { VisitBooking, VisitSlot, User } from '../data/schema';

export interface BookVisitDto {
  visitSlotId: number;
  animalId: number;
}

export class VisitService {
  /**
   * Book a visit slot for an animal.
   * CRITICAL PAWTRACK RULE: Booking a visit DOES NOT change Animal.Status!
   * The animal remains Available for other adopters.
   */
  public bookVisit(dto: BookVisitDto, currentUser: User): VisitBooking {
    const slot = dbContext.visitSlots.find((s) => s.id === dto.visitSlotId);
    if (!slot) {
      throw new Error(`Visit slot with ID ${dto.visitSlotId} was not found.`);
    }

    if (slot.bookedCount >= slot.capacity) {
      throw new Error('This visit slot has reached its maximum capacity.');
    }

    const animal = dbContext.animals.find((a) => a.id === dto.animalId);
    if (!animal) {
      throw new Error(`Animal with ID ${dto.animalId} was not found.`);
    }

    if (animal.branchId !== slot.branchId) {
      throw new Error('Selected visit slot belongs to a different branch than the animal.');
    }

    // Create booking
    const booking: VisitBooking = {
      id: dbContext.getNextId('visitBookings'),
      visitSlotId: dto.visitSlotId,
      animalId: dto.animalId,
      adopterId: currentUser.id,
      bookingDate: new Date().toISOString().split('T')[0],
      status: 'Booked'
    };

    // Increment slot booked count
    slot.bookedCount += 1;

    dbContext.visitBookings.push(booking);
    return booking;
  }

  /**
   * Update visit booking status (CheckedIn, Completed, Cancelled, NoShow)
   */
  public updateVisitStatus(bookingId: number, status: 'CheckedIn' | 'Completed' | 'Cancelled' | 'NoShow', currentUser: User): VisitBooking {
    const booking = dbContext.visitBookings.find((b) => b.id === bookingId);
    if (!booking) {
      throw new Error(`Visit booking with ID ${bookingId} was not found.`);
    }

    const animal = dbContext.animals.find((a) => a.id === booking.animalId);
    if (currentUser.role !== 'OrgAdmin' && animal && currentUser.branchId !== animal.branchId) {
      throw new Error('Access denied: Staff can only manage visits for their own branch.');
    }

    booking.status = status;
    return booking;
  }

  /**
   * Get available visit slots for a branch
   */
  public getVisitSlots(branchId: number): VisitSlot[] {
    return dbContext.visitSlots.filter((s) => s.branchId === branchId);
  }

  /**
   * Get visit bookings for currentUser or branch
   */
  public getVisitBookings(currentUser: User): any[] {
    let bookings = [...dbContext.visitBookings];

    if (currentUser.role === 'Adopter') {
      bookings = bookings.filter((b) => b.adopterId === currentUser.id);
    } else if (currentUser.role !== 'OrgAdmin' && currentUser.branchId !== null) {
      const branchAnimalIds = new Set(
        dbContext.animals.filter((a) => a.branchId === currentUser.branchId).map((a) => a.id)
      );
      bookings = bookings.filter((b) => branchAnimalIds.has(b.animalId));
    }

    return bookings.map((b) => {
      const animal = dbContext.animals.find((a) => a.id === b.animalId);
      const adopter = dbContext.users.find((u) => u.id === b.adopterId);
      const slot = dbContext.visitSlots.find((s) => s.id === b.visitSlotId);

      return {
        ...b,
        animalName: animal?.name || 'Unknown',
        animalSpecies: animal?.species || '',
        adopterName: adopter?.name || 'Unknown',
        slotDate: slot?.date,
        slotTime: slot ? `${slot.startTime} - ${slot.endTime}` : ''
      };
    });
  }
}

export const visitService = new VisitService();

File: Paw_Track-Anshu_IN26011396/src/PawTrack.Services/VisitService.cs
using Microsoft.EntityFrameworkCore;
using PawTrack.Core.DTOs;
using PawTrack.Core.Entities;
using PawTrack.Core.Enums;
using PawTrack.Core.Interfaces;
using PawTrack.Data;

namespace PawTrack.Services;

public class VisitService : IVisitService
{
    private readonly PawTrackDbContext _context;

    public VisitService(PawTrackDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Book a visit slot for an animal.
    /// CRITICAL PAWTRACK RULE: Booking a visit DOES NOT change Animal.Status!
    /// The animal remains Available for other adopters.
    /// </summary>
    public async Task<VisitBooking> BookVisitAsync(BookVisitDto dto, User currentUser)
    {
        var slot = await _context.VisitSlots.FirstOrDefaultAsync(s => s.Id == dto.VisitSlotId);
        if (slot == null)
        {
            throw new KeyNotFoundException($"Visit slot with ID {dto.VisitSlotId} was not found.");
        }

        if (slot.BookedCount >= slot.Capacity)
        {
            throw new InvalidOperationException("This visit slot has reached its maximum capacity.");
        }

        var animal = await _context.Animals.FirstOrDefaultAsync(a => a.Id == dto.AnimalId);
        if (animal == null)
        {
            throw new KeyNotFoundException($"Animal with ID {dto.AnimalId} was not found.");
        }

        if (animal.BranchId != slot.BranchId)
        {
            throw new InvalidOperationException("Selected visit slot belongs to a different branch than the animal.");
        }

        var booking = new VisitBooking
        {
            VisitSlotId = dto.VisitSlotId,
            AnimalId = dto.AnimalId,
            AdopterId = currentUser.Id,
            BookingDate = DateTime.UtcNow.Date,
            Status = VisitBookingStatus.Booked
        };

        slot.BookedCount += 1;

        _context.VisitBookings.Add(booking);
        await _context.SaveChangesAsync();

        return booking;
    }

    /// <summary>
    /// Update visit booking status (CheckedIn, Completed, Cancelled, NoShow)
    /// </summary>
    public async Task<VisitBooking> UpdateVisitStatusAsync(int bookingId, VisitBookingStatus status, User currentUser)
    {
        var booking = await _context.VisitBookings
            .Include(b => b.Animal)
            .FirstOrDefaultAsync(b => b.Id == bookingId);

        if (booking == null)
        {
            throw new KeyNotFoundException($"Visit booking with ID {bookingId} was not found.");
        }

        if (currentUser.Role != UserRole.OrgAdmin && booking.Animal != null && currentUser.BranchId != booking.Animal.BranchId)
        {
            throw new UnauthorizedAccessException("Access denied: Staff can only manage visits for their own branch.");
        }

        booking.Status = status;
        await _context.SaveChangesAsync();

        return booking;
    }

    /// <summary>
    /// Get available visit slots for a branch
    /// </summary>
    public async Task<List<VisitSlotResponseDto>> GetVisitSlotsAsync(int branchId)
    {
        var slots = await _context.VisitSlots
            .Where(s => s.BranchId == branchId)
            .OrderBy(s => s.Date)
            .ThenBy(s => s.StartTime)
            .ToListAsync();

        return slots.Select(s => new VisitSlotResponseDto
        {
            Id = s.Id,
            BranchId = s.BranchId,
            Date = s.Date,
            StartTime = s.StartTime,
            EndTime = s.EndTime,
            Capacity = s.Capacity,
            BookedCount = s.BookedCount
        }).ToList();
    }

    /// <summary>
    /// Get visit bookings for currentUser or branch
    /// </summary>
    public async Task<List<VisitBookingResponseDto>> GetVisitBookingsAsync(User currentUser)
    {
        var query = _context.VisitBookings
            .Include(b => b.Animal)
            .Include(b => b.Adopter)
            .Include(b => b.VisitSlot)
            .AsQueryable();

        if (currentUser.Role == UserRole.Adopter)
        {
            query = query.Where(b => b.AdopterId == currentUser.Id);
        }
        else if (currentUser.Role != UserRole.OrgAdmin && currentUser.BranchId.HasValue)
        {
            query = query.Where(b => b.Animal != null && b.Animal.BranchId == currentUser.BranchId.Value);
        }

        var bookings = await query.OrderByDescending(b => b.BookingDate).ToListAsync();

        return bookings.Select(b => new VisitBookingResponseDto
        {
            Id = b.Id,
            VisitSlotId = b.VisitSlotId,
            AnimalId = b.AnimalId,
            AnimalName = b.Animal?.Name ?? "Unknown",
            AnimalSpecies = b.Animal?.Species ?? string.Empty,
            AdopterId = b.AdopterId,
            AdopterName = b.Adopter?.Name ?? "Unknown",
            BookingDate = b.BookingDate,
            SlotDate = b.VisitSlot?.Date,
            SlotTime = b.VisitSlot != null
                ? $"{b.VisitSlot.StartTime:hh\\:mm} - {b.VisitSlot.EndTime:hh\\:mm}"
                : string.Empty,
            Status = b.Status,
            StaffNotes = b.StaffNotes
        }).ToList();
    }
}
